src/utils/map.hpp

```diff
@@ -0,0 +1,968 @@
+#pragma once
+#include <functional>
+#include <iostream>
+#include <iomanip>
+
+
+
+namespace nstd
+{
+
+    /* FORWARD DECLARATIONS*/
+
+    template <typename key>
+    class hash;
+
+    template <typename key, typename value, typename Hash>
+    class map;
+
+    template <class k, class v, class Hash>
+    std::ostream& operator<<(std::ostream& out, const map<k, v, Hash>& d);
+
+    unsigned int* SieveOfEratosthenes(int n);
+
+    /*CLASSES & STRUCTS */
+
+    /**
+     * @class nstd::map class
+     *
+     * A data structure class that is meant to hold key value pairs and
+     * hash with the key to retrieve the value. This map is also iteratable
+     * meaning that you can start at the beginning or end and read through it
+     * as if it was a list. In the context of our project this class will relate
+     * team id's with a stadium and stadium id's with a team.
+     *
+     * This class also supports custom hashing, but the functions need to be in this
+     * format static int staticHashAlgo(key k,int j,int capacity) where the returned
+     * int from the function needs to follow this restriction 0 < hash < capacity.
+     *
+     * @code{.cpp}
```

Comment:
This code example is a little compact. It also doesn't show reassigning values
```cpp
nstd::map<int,std::string> myMap;

//Create a new entry in the map
myMap[2] = "Hello";

//Reassign value at key 2
myMap[2] = "world!";

//Print out value at key 2
std::cout << myMap[2];

//Prints out value's default constructor since the key doesn't exist
std::cout << myMap[8];
```